crates/protocol/src/block_info.rs

```diff
@@ -566,6 +614,70 @@ impl L1BlockInfoEcotone {
             base_fee_scalar,
         })
     }
+
+    /// Calculate the data gas for posting the transaction on L1. Calldata costs 16 gas per byte
+    /// after compression.
+    ///
+    /// Prior to fjord, calldata costs 16 gas per non-zero byte and 4 gas per zero byte.
+    pub fn data_gas(&self, input: &[u8]) -> U256 {
+        let rollup_data_gas_cost = U256::from(input.iter().fold(0, |acc, byte| {
+            acc + if *byte == 0x00 { ZERO_BYTE_COST } else { NON_ZERO_BYTE_COST }
+        }));
+
+        rollup_data_gas_cost
+    }
+
+    /// Calculate the gas cost of a transaction based on L1 block data posted on L2, post-Ecotone.
+    ///
+    /// [SpecId::ECOTONE] L1 cost function:
+    /// `(calldataGas/16)*(l1BaseFee*16*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar)/1e6`
+    ///
+    /// We divide "calldataGas" by 16 to change from units of calldata gas to "estimated # of bytes when compressed".
+    /// Known as "compressedTxSize" in the spec.
+    ///
+    /// Function is actually computed as follows for better precision under integer arithmetic:
+    /// `calldataGas*(l1BaseFee*16*l1BaseFeeScalar + l1BlobBaseFee*l1BlobBaseFeeScalar)/16e6`
+    pub fn calculate_tx_l1_cost(&self, input: &[u8], empty_scalers: bool) -> U256 {
+        if input.is_empty() || input.first() == Some(&0x7F) {
+            return U256::ZERO;
+        }
+
+        // There is an edgecase where, for the very first Ecotone block (unless it is activated at Genesis), we must
+        // use the Bedrock cost function. To determine if this is the case, we can check if the Ecotone parameters are
+        // unset.
+        if empty_scalers {
+            return self.calculate_tx_l1_cost_bedrock(input);
+        }
+
+        let rollup_data_gas_cost = self.data_gas(input);
+        let l1_fee_scaled = self.calculate_l1_fee_scaled_ecotone();
+
+        l1_fee_scaled
+            .saturating_mul(rollup_data_gas_cost)
+            .wrapping_div(U256::from(1_000_000 * NON_ZERO_BYTE_COST))
+    }
+
+    /// Calculate the gas cost of a transaction based on L1 block data posted on L2, pre-Ecotone.
+    pub fn calculate_tx_l1_cost_bedrock(&self, input: &[u8]) -> U256 {
+        let rollup_data_gas_cost = self.data_gas(input);
+        rollup_data_gas_cost
+            //.saturating_add(self.l1_fee_overhead)
```

Comment:
Not very sure what's the best way to deal with `l1_fee_overhead` for post-`ecotone` blocks, is it worth to add the attribute to the struct?

Reply:
This function should only ever be used on a `L1BlockInfoBedrock` variant. When the type is decoded from calldata or storage, the variant is decided. The fee functions on each variant need to support the forks that span their activity, i.e. `L1BlockInfoEcotone` needs to support ecotone + fjord changes, and `L1BlockInfoBedrock` needs to support bedrock + regolith changes.

At the time that we decode them in `revm`, we'll decide the same way we do today, with checks like the [empty scalars check](https://github.com/bluealloy/revm/blob/ca5b2fe15e0b9920ede13f1c67918b93bfdc12c2/crates/optimism/src/l1block.rs#L108-L115). When decoding from calldata, we can switch on the function selector.